# Part 1- 7 F1 racers go to 7 shows where each racer goes to one and only one show for 7 minutes

We first create a sample space where the columns demonstrate the fashion show number different drivers visited. It is a list of tuples. 
For example ("1", "4", "4", "4", "2", "3", "7") means Driver 1 visited the first fashion show, Drivers 2, 3, and 4 visited the fourth fashion show. Driver 5 the third and Driver 7 the seventh fashion show. 


In [1]:
from itertools import product
N = 7 
res = []
for ele in product(range(1, N + 1), repeat = N):
    res.append(ele)

In [2]:
len(res)
res_set = set(res) # to check if there are duplicates and verify our sample space. 
print(len(res) == len(res_set))
res[:10]

True


[(1, 1, 1, 1, 1, 1, 1),
 (1, 1, 1, 1, 1, 1, 2),
 (1, 1, 1, 1, 1, 1, 3),
 (1, 1, 1, 1, 1, 1, 4),
 (1, 1, 1, 1, 1, 1, 5),
 (1, 1, 1, 1, 1, 1, 6),
 (1, 1, 1, 1, 1, 1, 7),
 (1, 1, 1, 1, 1, 2, 1),
 (1, 1, 1, 1, 1, 2, 2),
 (1, 1, 1, 1, 1, 2, 3)]

In [3]:
#probability of seeing driver if he goes to first fashion show. 
count = 0 
for i in res:
    if (1 in i):
        count = count + 1 
        
p = count/len(res)
p


0.6600833229108862

The probability of going to fashion 1: P(F1) = 1/7
The probabilty of seeing a driver at fashion show 1 = 0.6600833229108862

total probability = P(F1) + P(F2) + P(F3) + P(F4) + P(F5) + P(F6) + P(F7) = (1/7*p) * 7 = p
                        

# Part 2 one F1 driver goes to one or more of 7 shows for exactly 7 times and for 7 minutes each time

No of combinations of 1 F1 Racer attending 7 shows is equal to 7 racers attending 7 shows. 
For this condition, the time slots are like 7 different drivers hence the sample space is exactly same where the columns represent timeslots. For example 
("1", "4", "4", "4", "2", "3", "7") means that the driver goes to Fashion Show-1 in the first time slot, to Fashion show 4 in the next 3 time slots and attends Fashion Show 2, 3, and 7 for the next 3 time slots. 

Hence the sample space will be the same as before which is list **res**

The probability of going to fashion 1: P(F1) = 1/7
The probabilty of seeing a driver at fashion show 1 = 0.6600833229108862

total probability = P(F1) + P(F2) + P(F3) + P(F4) + P(F5) + P(F6) + P(F7) = (1/7*p) * 7 = p

In [4]:
#probability of seeing driver if he goes to first fashion show. 
count = 0 
for i in res:
    if (1 in i):
        count = count + 1 
        
p_one_driver_multiple_shows = count/len(res_set)
p_one_driver_multiple_shows

0.6600833229108862

The probability of going to fashion 1: P(F1) = 1/7
The probabilty of seeing a driver at fashion show 1 = 0.6600833229108862

total probability = P(F1) + P(F2) + P(F3) + P(F4) + P(F5) + P(F6) + P(F7) = (1/7*p) * 7 = p

In [5]:
((1/7)*p_one_driver_multiple_shows)*7

0.6600833229108862

Alternatively, after a word with the TAs, and exploring the clues together we have come up with another approach to solve the problem disregrading the order in the combinations generated and we found something interesting. 

In this case only the number of times a fashion show was visited by the driver matters, the order does not. 

We first generate the sample space, find the probability if Professor visits the first fashion show and then add the probabilities in the same way as above using the following:


The probability of going to fashion 1: P(F1) = 1/7 The probabilty of seeing a driver at fashion show 1 = 0.6600833229108862

total probability = P(F1) + P(F2) + P(F3) + P(F4) + P(F5) + P(F6) + P(F7) = (1/7*p) * 7 = p


In [6]:
from itertools import combinations_with_replacement

fashion_shows = '1234567'
sample_no_order = list(combinations_with_replacement(fashion_shows, 7))
sample_no_order[:100]

[('1', '1', '1', '1', '1', '1', '1'),
 ('1', '1', '1', '1', '1', '1', '2'),
 ('1', '1', '1', '1', '1', '1', '3'),
 ('1', '1', '1', '1', '1', '1', '4'),
 ('1', '1', '1', '1', '1', '1', '5'),
 ('1', '1', '1', '1', '1', '1', '6'),
 ('1', '1', '1', '1', '1', '1', '7'),
 ('1', '1', '1', '1', '1', '2', '2'),
 ('1', '1', '1', '1', '1', '2', '3'),
 ('1', '1', '1', '1', '1', '2', '4'),
 ('1', '1', '1', '1', '1', '2', '5'),
 ('1', '1', '1', '1', '1', '2', '6'),
 ('1', '1', '1', '1', '1', '2', '7'),
 ('1', '1', '1', '1', '1', '3', '3'),
 ('1', '1', '1', '1', '1', '3', '4'),
 ('1', '1', '1', '1', '1', '3', '5'),
 ('1', '1', '1', '1', '1', '3', '6'),
 ('1', '1', '1', '1', '1', '3', '7'),
 ('1', '1', '1', '1', '1', '4', '4'),
 ('1', '1', '1', '1', '1', '4', '5'),
 ('1', '1', '1', '1', '1', '4', '6'),
 ('1', '1', '1', '1', '1', '4', '7'),
 ('1', '1', '1', '1', '1', '5', '5'),
 ('1', '1', '1', '1', '1', '5', '6'),
 ('1', '1', '1', '1', '1', '5', '7'),
 ('1', '1', '1', '1', '1', '6', '6'),
 ('1', '1', 

In [7]:
count = 0 
for i in sample_no_order:
    if ("1" in i):
        count = count + 1 
        
p = count/len(sample_no_order)
p

0.5384615384615384

# Solving it Using the famous **p** function 

In [8]:
from fractions import Fraction
class ProbDist(dict):
    """A Probability Distribution; an {outcome: probability} mapping."""
    def __init__(self, mapping=(), **kwargs):
        self.update(mapping, **kwargs)
        # Make probabilities sum to 1.0; assert no negative probabilities
        total = sum(self.values())
        for outcome in self:
            self[outcome] = self[outcome] / total
            assert self[outcome] >= 0

def joint(A, B, sep=' '):
    """The joint distribution of two independent probability distributions. 
    Result is all entries of the form {a+sep+b: P(a)*P(b)}"""
    return ProbDist({a + sep + b: A[a] * B[b]
                    for a in A
                    for b in B})

def p(event, space): 
    """The probability of an event, given a sample space of equiprobable outcomes. 
    event: a collection of outcomes, or a predicate that is true of outcomes in the event. 
    space: a set of outcomes or a probability distribution of {outcome: frequency} pairs."""
    # branch on the type of the first argument
    if is_predicate(event):
        # transform the mapping (untangible) 'event' into the collection (tangible) 'event'
        event = such_that(event, space)
        
    if isinstance(space, ProbDist):
        # if space is a dictionary of distinct probabilities, where each item does not count as the same amount
        return sum(space[k] for k in space if k in event)
        #return sum({space[k] for k in space & event})
    else:
        # space is not a dictionary but a collection, let's fall back to our original division
        return Fraction(len(event & space), len(space))

is_predicate = callable

def such_that(predicate, space): 
    """The outcomes in the sample pace for which the predicate is true.
    If space is a set, return a subset {outcome,...} with outcomes where predicate(element) is true;
    if space is a ProbDist, return a ProbDist {outcome: frequency,...} with outcomes where predicate(element) is true."""
    if isinstance(space, ProbDist):
        return ProbDist({o:space[o] for o in space if predicate(o)})
    else:
        return {o for o in space if predicate(o)}


In [9]:
#predicates for driver visiting the first fashion show. 

def not_alone(outcome) : return 1 in outcome #if fashion show 1 gets a visit by the driver

p_prof_meets_in_show_1 = p(not_alone, res_set)
p_prof_meets = ((1/7)*p_prof_meets_in_show_1)*7
p_prof_meets

0.6600833229108862

In [10]:
#it can be seen that the probability is equal for other fashion shows as well
#predicates for driver visiting the first fashion show. 

def not_alone(outcome) : return 2 in outcome #

p_prof_meets_in_show_2 = p(not_alone, res_set)
p_prof_meets_in_show_2

Fraction(543607, 823543)

# Alternate theory using p function where the order of visit does not matter 

In [11]:
def not_alone_alternate(outcome) : return "1" in outcome #defining another predicate because the other sample space has string values. 

sample_no_order_set = set(sample_no_order)
p_prof_meets_in_show_1 = p(not_alone_alternate, sample_no_order_set)
p_prof_meets = ((1/7)*p_prof_meets_in_show_1)*7
p_prof_meets

0.5384615384615384

It can be seen that our answers match for both the methods. 

# Using Monte Carlo

In [13]:
import numpy as np
from tqdm import tqdm



numdrivers = 7
numshows = 7
def sim(n):
    yes = 0
    no = 0
    
    for i in tqdm(range(n)):
        shows = []
        shows.extend(np.random.randint(low = 0,high = 7, size=7))
        me = np.random.randint(low = 0,high = 6, size=1)
        
        if me in shows:
            yes+=1
        else:
            no+=1
            
    return yes / n

In [14]:
sim(1000)

100%|███████████████████████████████████████████████████████████████████████████| 1000/1000 [00:00<00:00, 66693.76it/s]


0.647